contracts/token-registrars/CanonicalTokenRegistrar.sol

```diff
@@ -0,0 +1,98 @@
+import { ITokenManagerType } from '../interfaces/ITokenManagerType.sol';
+import { IERC20Named } from '../interfaces/IERC20Named.sol';
+
+import { Multicall } from '../utils/Multicall.sol';
+
+contract CanonicalTokenRegistrar is ICanonicalTokenRegistrar, ITokenManagerType, Multicall, Upgradable {
+    using SafeTokenTransferFrom for IERC20;
+
+    IInterchainTokenService public immutable service;
+    bytes32 public immutable chainNameHash;
+
+    bytes32 internal constant PREFIX_CANONICAL_TOKEN_SALT = keccak256('canonical-token-salt');
+    bytes32 private constant CONTRACT_ID = keccak256('canonical-token-registrar');
+
+    constructor(address interchainTokenServiceAddress) {
+        if (interchainTokenServiceAddress == address(0)) revert ZeroAddress();
+        service = IInterchainTokenService(interchainTokenServiceAddress);
+        string memory chainName_ = IInterchainTokenService(interchainTokenServiceAddress).remoteAddressValidator().chainName();
+        chainNameHash = keccak256(bytes(chainName_));
+    }
+
+    /**
+     * @notice Getter for the contract id.
+     */
+    function contractId() external pure returns (bytes32) {
+        return CONTRACT_ID;
+    }
+
+    function getCanonicalTokenSalt(address tokenAddress) public view returns (bytes32 salt) {
+        salt = keccak256(abi.encode(PREFIX_CANONICAL_TOKEN_SALT, chainNameHash, tokenAddress));
+    }
+
+    function getCanonicalTokenId(address tokenAddress) public view returns (bytes32 tokenId) {
+        tokenId = service.getCustomTokenId(address(this), getCanonicalTokenSalt(tokenAddress));
+    }
+
+    function registerCanonicalToken(address tokenAddress) external payable returns (bytes32 tokenId) {
+        bytes memory params = abi.encode('', tokenAddress);
+        bytes32 salt = getCanonicalTokenSalt(tokenAddress);
+        tokenId = service.deployCustomTokenManager(salt, TokenManagerType.LOCK_UNLOCK, params);
+    }
+
+    function deployAndRegisterRemoteCanonicalToken(bytes32 salt, string calldata destinationChain, uint256 gasValue) external payable {
+        // This ensures that the token manages has been deployed by this address, so it's safe to trust it.
+        bytes32 tokenId = service.getCustomTokenId(address(this), salt);
+        IERC20Named token = IERC20Named(service.getTokenAddress(tokenId));
+        // The 3 lines below will revert if the token manager does not exist.
+        string memory tokenName = token.name();
+        string memory tokenSymbol = token.symbol();
+        uint8 tokenDecimals = token.decimals();
+
+        // slither-disable-next-line arbitrary-send-eth
+        service.deployAndRegisterRemoteStandardizedToken{ value: gasValue }(
+            salt,
+            tokenName,
+            tokenSymbol,
+            tokenDecimals,
+            '',
+            '',
+            0,
+            '',
+            destinationChain,
+            gasValue
+        );
+    }
+
+    function transferCanonicalToken(
```

Comment:
If we can figure out how to make this generic for other token types, then we could drop the initial mint amount from the official ITS deploy payload spec altogether.

Perhaps this approach could work:
- The deploy method for the current chain takes in mintTo/mintAmount argument.
- The deploy remote method doesn't forward mintAmount anymore. It just deploys.
- The UI sets the mintTo address to be the registrar contract.
- The subsequent transfer calls in the multicall then call interchainTransfer method (with an approval to the token manager if necessary).
- This way the user doesn't need to pre-approve a non-existent token.